pkg/controller/cluster/cluster_controller.go

```diff
@@ -356,7 +366,96 @@ func (c *Controller) processNextWorkItem() bool {
 	return true
 }
 
+func (c *Controller) buildRemoteClusterClient(cluster *clusteroperator.Cluster) (clusterapiclient.Interface, error) {
+
+	secretName := cluster.Name + "-kubeconfig"
+
+	secret, err := c.kubeClient.CoreV1().Secrets(cluster.Namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	secretData := secret.Data["kubeconfig"]
+
+	config, err := clientcmd.Load(secretData)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeConfig := clientcmd.NewDefaultClientConfig(*config, &clientcmd.ConfigOverrides{})
+	restConfig, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	clusterAPIClient, err := clusterapiclient.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	return clusterAPIClient, nil
+}
+
+func (c *Controller) setDeprovisionedComputeMachinesets(cluster *clusteroperator.Cluster) error {
+	original := cluster
+	cluster = cluster.DeepCopy()
+	cluster.Status.DeprovisionedComputeMachinesets = true
+	return controller.PatchClusterStatus(c.client, original, cluster)
+}
+
+func (c *Controller) ensureRemoteMachineSetsAreDeleted(cluster *clusteroperator.Cluster) error {
+	clusterAPIClient, err := c.buildRemoteClusterClient(cluster)
+	if err != nil {
+		return err
+	}
+	machineSets, err := clusterAPIClient.ClusterV1alpha1().MachineSets(remoteClusterNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	// If all remote machinesets are gone, then we should update the cluster status
+	if len(machineSets.Items) == 0 {
+		return c.setDeprovisionedComputeMachinesets(cluster)
+	}
+
+	if err = clusterAPIClient.ClusterV1alpha1().MachineSets(remoteClusterNamespace).DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{}); err != nil {
+		return err
+	}
+
+	// Ensure that the cluster is enqueued again so we can continue to monitor machinesets
+	c.enqueueAfter(cluster, deleteRemoteInterval)
+	return nil
+}
+
+func (c *Controller) shouldDeleteRemoteMachineSets(cluster *clusteroperator.Cluster) (bool, error) {
+	if cluster.Status.DeprovisionedComputeMachinesets {
+		return false, nil
+	}
+	if len(cluster.Status.MasterMachineSetName) == 0 {
+		return false, nil
+	}
+	masterMachineSet, err := c.machineSetsLister.MachineSets(cluster.Namespace).Get(cluster.Status.MasterMachineSetName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	if !masterMachineSet.Status.ClusterAPIInstalled {
+		return false, nil
+	}
+	return true, nil
+}
+
 func (c *Controller) ensureMachineSetsAreDeleted(cluster *clusteroperator.Cluster) error {
+	shouldDeleteRemote, err := c.shouldDeleteRemoteMachineSets(cluster)
+	if err != nil {
+		return err
+	}
+	if shouldDeleteRemote {
+		return c.ensureRemoteMachineSetsAreDeleted(cluster)
+	}
+	if !cluster.Status.DeprovisionedComputeMachinesets {
```

Comment:
I am concerned about the potential for a race condition here where other controllers may be in the process of installing the cluster-api and provisioning the compute machinesets in the remote cluster.

Reply:
That wouldn't be sufficient. That would only decrease the window of the race condition. There is still the chance that this controller uses a revision of the Cluster that does not have the status updated to reflect that the cluster-api is installing. We could avoid this if we did not patch and retry on conflicts.

Reply:
We talked before about not doing the patch/retry thing, but came back to the conclusion that it could cause more problems having several controllers trying to update the same resource. Something that maybe we could explore doing is having at most one controller per resource. Either change I think would be beyond the scope of this PR :) Would you be ok with creating an issue and looking into fixing it separately?

Reply:
Could a fix be to move this logic to the controller that provisions the machinesets on to the remote cluster? We would be guaranteed that we're not trying to delete the remote machinesets at the same time that we're provisioning them.

Reply:
In reconsidering this, I am fine with kicking this can down the road in the interest of pushing forward the migration to using the upstream API fully.